src/smc-util/upgrades/quota.ts

```diff
@@ -199,233 +238,346 @@ function calc_default_quotas(site_settings?: SiteSettingsQuotas): Quota {
+    "0-0": [],
+    "0-1": [],
+    "1-0": [],
+    "1-1": [],
+  };
 
-  // site settings max quotas overwrite the hardcoded values
-  const max_upgrades = Object.assign(
-    {},
-    MAX_UPGRADES,
-    site_settings?.max_upgrades ?? {}
-  );
+  // classification
+  for (const [key, val] of Object.entries(site_licenses)) {
+    const is_ar = isSiteLicenseQuotaSetting(val)
+      ? val.quota.always_running === true
+      : (val.always_running ?? 0) >= 1;
 
-  // network access
-  if (max_upgrades.network == 0) {
-    quota.network = false;
-  } else if (!quota.network) {
-    if (settings.network) {
-      // free admin-set
-      quota.network = true;
-    } else {
-      // paid by some user
-      for (const userid in users) {
-        const val = users[userid];
-        if (val != null && val.upgrades && val.upgrades.network) {
-          quota.network = true;
-          break;
-        }
-      }
-      // or some site license
-      if (!quota.network && site_license != null) {
-        for (const license_id in site_license) {
-          const val = site_license[license_id];
-          if (val != null && val.network) {
-            quota.network = true;
-            break;
-          }
-        }
-      }
-    }
+    const is_member = isSiteLicenseQuotaSetting(val)
+      ? val.quota.member === true
+      : (val.member_host ?? 0) >= 1;
+
+    groups[`${is_member ? "1" : "0"}-${is_ar ? "1" : "0"}`].push(key);
   }
 
-  // member hosting, which translates to better hosting conditions of the project
-  if (max_upgrades.member_host == 0) {
-    quota.member_host = false;
-  } else if (!quota.member_host) {
-    if (settings.member_host) {
-      // free admin-set
-      quota.member_host = true;
-    } else {
-      // paid by some user
-      for (const userid in users) {
-        const val = users[userid];
-        if (val != null && val.upgrades && val.upgrades.member_host) {
-          quota.member_host = true;
-          break;
+  // selection -- always_running comes first, then member hosting
+  const selected = (function () {
+    for (const ar of ["1", "0"]) {
+      // always running
+      for (const mh of ["1", "0"]) {
+        // member hosting
+        const k = `${mh}-${ar}`;
+        if (groups[k].length > 0) {
+          return groups[k];
         }
       }
-      // or some site license
-      if (!quota.member_host && site_license != null) {
-        for (const license_id in site_license) {
-          const val = site_license[license_id];
-          if (val != null && val.member_host) {
-            quota.member_host = true;
-            break;
-          }
-        }
+    }
+  })();
+
+  return selected.reduce((acc, cur) => {
+    acc[cur] = site_licenses[cur];
+    return acc;
+  }, {});
+}
+
+// there is an old schema, inherited from SageMathCloud, etc. and newer iterations.
+// this helps by going from one schema to the newer one
+function upgrade2quota(up: Partial<Upgrades>): RQuota {
+  const dflt_false = (x) =>
```

Comment:
Sorry but it is pretty funny that you abbreviated "default" to "dflt" (I assume).